File: 2022/04/solution.hs
import Data.List.Split (splitOn)

data Interval = I Int Int deriving Show

parseInterval :: String -> Interval
parseInterval s = (I low high) where
  splitted = splitOn "-" s
  (low,high) = case splitted of
    (s1:s2:[]) -> ((read s1 :: Int), (read s2 :: Int))
    _ -> error "Malformatted interval"

parseLine :: String -> (Interval, Interval)
parseLine s = (i1, i2) where
  splitted = splitOn "," s
  (i1,i2) = case splitted of
    (s1:s2:[]) -> (parseInterval s1, parseInterval s2)
    _ -> error "Malformatted line"

intervalContains :: Interval -> Interval -> Bool
intervalContains (I l1 h1) (I l2 h2) = l1 <= l2 && h2 <= h1

intervalsOverlap :: Interval -> Interval -> Bool
intervalsOverlap (I l1 h1) (I l2 h2) = not (h2 < l1 || l2 > h1)

part :: (Interval -> Interval -> Bool) -> [(Interval, Interval)] -> Int
part fn ss = sum where
  helper (i1, i2) = fromEnum (fn i1 i2)
  tmp = map helper ss
  sum = foldl (+) 0 tmp

part1 :: [(Interval, Interval)] -> Int
part1 ss = part (\i1 i2 -> intervalContains i1 i2 || intervalContains i2 i1) ss

part2 :: [(Interval, Interval)] -> Int
part2 ss = part intervalsOverlap ss

main = do
  let fileName = "input.dat"
  input <- readFile fileName
  let input_lines = lines input
  let intervals = map parseLine input_lines
  print (part1 intervals)
  print (part2 intervals)

File: 2022/04/solution.py
import pathlib
from dataclasses import dataclass


def part1(entries):
    return sum(i1.contains(i2) or i2.contains(i1) for i1, i2 in entries)


def part2(entries):
    return sum(i1.overlaps(i2) for i1, i2 in entries)


@dataclass
class Interval:
    low: int
    high: int

    def overlaps(self, other):
        return not (other.high < self.low or other.low > self.high)

    def contains(self, other):
        return self.low <= other.low and other.high <= self.high


def main():
    input = pathlib.Path("input.dat")

    entries = []
    for l in input.read_text().split("\n"):
        if not l:
            continue
        pair = l.split(",")
        iv1 = pair[0].split("-")
        iv2 = pair[1].split("-")

        interval1 = Interval(int(iv1[0]), int(iv1[1]))
        interval2 = Interval(int(iv2[0]), int(iv2[1]))
        entries.append((interval1, interval2))
    print(part1(entries))
    print(part2(entries))


if __name__ == "__main__":
    main()
